viite-backend/viite-main/src/main/scala/fi/liikennevirasto/viite/ProjectService.scala

```diff
@@ -1580,45 +1580,71 @@ def setCalibrationPoints(startCp: Long, endCp: Long, projectLinks: Seq[ProjectLi
         logger.warn(s"Ely-codes for project: $projectId were not updated.")
     }
 
+
+    /**
+     * Updates project links with data from the dev tool.
+     *
+     * This function takes the manually given devToolData, which contains the new address, roadway, and side code information,
+     * and updates the corresponding project links with these values. It also handles the creation of new
+     * roadway numbers if necessary, adjusts status and side codes, and sets calibration points for roadways.
+     *
+     * The function performs the following steps:
+     * 1. If both start and end address M values are defined, it updates the project links with these address values.
+     * 2. If none of the project links are marked as 'New', it updates the original address values.
+     * 3. If all project links are new, it generates a new roadway number for those links.
+     * 4. Updates the status and side code for the project links, adjusting them as necessary based on the change type.
+     * 5. Retrieves the original road addresses corresponding to the updated project links using their roadway IDs.
+     * 6. Sets calibration points based on the provided start and end calibration points.
+     * 7. Optionally generates a new roadway number for all updated links if required by the devToolData.
+     * 8. If an edited side code is provided, it updates the side codes for the project links.
+     * 9. Updates the project links in the database using the `projectLinkDAO.updateProjectLinks` method.
```

Comment:
Hmm. In fact.
Function having this many things to do, could be sliced into a smaller functions.
And in this function, there should be only calls to the functions each performing stuff at a single numbered line.

Reply:
Jees, tämä oli hyvä.
Nyt kunkin palan tajuaa kerralla, ja ymmärrettävää on aina vain yks pala kerrallan. :)
💪